### PR TITLE
Issue433: parsing Javadoc comments

### LINKED ISSUE #433: Parsing JavaDoc
If possible to improve the parsing of a javaDoc comment Block, by adding javaDoc annotation like @return, @param ... .
I think is the last missing things to javaparser.
In the same idea, add some methode to the Generic comment like  isTodoComment() and method like that.
I will try to create a little POC if you think this feature is interesting

### PR DESCRIPTION
**Pull request by ftomassetti**
We parse Javadoc comments to produce a `JavadocDocument`.

A `JavadocDocument` is composed by a `JavadocDescription` and a list of `JavadocBlockTags` (e.g. `@param myParam amazing param`).

A `JavadocDescription` is a list of text blocks and `JavadocInlineTags` (e.g., `{@see foo}`)

We have a list of well-known block and line tags and we try to recognize, however the user can also use arbitrary tags (type `UNKNOWN`).

Fix #433

### CODE REVIEW

**Comment by ftomassetti** [comment]
Mmm + 3.9% to coverage?

**Comment by matozoid** [comment]
Nice. I have to take some time to look at this carefully though.

**Comment by ftomassetti** [comment]
Yes, it would benefit from a good revision and probably also some additional tests

File: javaparser-core/src/main/java/com/github/javaparser/javadoc/JavadocDocument.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    public JavadocDocument addBlockTag(String tagName) {
+        return addBlockTag(tagName, "");
+    }
+
+    /**
+     * Return the text content of the document. It does not containing trailing spaces and asteriks
+     * at the start of the line.
+     */
+    public String toText() {
+        StringBuffer sb = new StringBuffer();
```
Comment (arturbosch) [inline javaparser-core/src/main/java/com/github/javaparser/javadoc/JavadocDocument.java]: A StringBuffer is synchronized, was this by a mistake? (I just scanned the modified files...)

File: javaparser-core/src/main/java/com/github/javaparser/javadoc/description/JavadocDescription.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    public JavadocDescription() {
+        elements = new LinkedList<>();
+    }
+
+    public void addElement(JavadocDescriptionElement element) {
+        this.elements.add(element);
+    }
+
+    public String toText() {
+        StringBuffer sb = new StringBuffer();
```
Comment (arturbosch) [inline javaparser-core/src/main/java/com/github/javaparser/javadoc/description/JavadocDescription.java]: Replace with StringBuilder here too

File: javaparser-core/src/main/java/com/github/javaparser/javadoc/JavadocDocument.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/javadoc/JavadocDocument.java]: Right, will fix. Thanks!

**Comment by matozoid** [comment]
Alright, mostly hygiene feedback:
- please give every class a Javadoc that gives the user an idea what its use is. If it represents some kind of text (like JavadocBlockTag) show a little example. Just enough to trigger "oh, that's what you mean!"
- the StringBuffer remark from @arturbosch 
- I think we can create a new Utils method that converts CONSTANT_TEXT to constantText for use in our enums
- `JavadocDocument` is a silly name :) What about simply `Javadoc`? It seems to be a central class for everything Javadoc.
- how about NodeWithJavadoc - maybe juggle the methods there to have `getJavadocRaw()` `getJavadocParsed()` `setJavadoc(String)` and `setJavadoc(JavadocDocument)` ? Not so sure about the naming...

**Comment by matozoid** [comment]
Another one:
* To make JavadocParser "equal" to JavaParser, it could reside in the same package, and offer a static `parse` method. And/or JavaParser can have a `parseJavadoc(String)` method.

**Comment by ftomassetti** [comment]
Implement `setJavadoc(JavadocDocument)` would not be trivial because to convert JavadocDocument to a string we would need to know the proper indentation, so it would leave this out of this PR

**Comment by ftomassetti** [comment]
Thank you to all of you for the comments. I tried to address them and rebase the PR on master. When updating `NodeWithJavadoc` I also tried to normalize `Javadoc` to be written always as `Javadoc` and not `JavaDoc` some times

**Comment by matozoid** [comment]
`setJavadoc(String indentation, Javadoc)` - only half serious this ;)

I am slightly bothered by not having a way to put a Javadoc back in the AST though...

**Comment by ftomassetti** [comment]
I will add `setJavadoc(String indentation, Javadoc)`. Then as a followup we can write a `setJavadoc(Javadoc)` which gets the indentation from the context of the node.
Would that work for you guys?

**Comment by matozoid** [comment]
+1 from me!

**Comment by ftomassetti** [comment]
Good, then I will rebase it on master and wait for the tests to pass before merging

**Comment by arturbosch** [comment]
Looks awesome! +1

**Comment by ftomassetti** [comment]
thanks!

**Comment by matozoid** [comment]
See #325

**Pull request merged**